pkg/relayer/proxy/relay_meter.go

```diff
@@ -0,0 +1,384 @@
+	servicetypes "github.com/pokt-network/poktroll/x/service/types"
+	sessionkeeper "github.com/pokt-network/poktroll/x/session/keeper"
+	sessiontypes "github.com/pokt-network/poktroll/x/session/types"
+	sharedtypes "github.com/pokt-network/poktroll/x/shared/types"
+)
+
+var _ relayer.RelayMeter = (*ProxyRelayMeter)(nil)
+
+// appRelayMeter is the relay meter's internal representation of an onchain
+// Application's max and consumed stake.
+type appRelayMeter struct {
+	// The onchain application the relay meter is for.
+	app apptypes.Application
+	// The maximum uPOKT an application can pay this relayer for a given session.
+	// This is a fraction of the Application's overall stake.
+	maxCoin cosmostypes.Coin
+	// The amount of uPOKT a specific application has consumed from this relayer in the given session.
+	consumedCoin cosmostypes.Coin
+	// The header for the session the Application and Supplier (backed by the relayer)
+	// are exchanging services in.
+	sessionHeader *sessiontypes.SessionHeader
+
+	// sharedParams, service and serviceRelayDifficulty are used to calculate the relay cost
+	// that increments the consumedAmount.
+	// They are cached at each session to avoid querying the blockchain for each relay.
+	// TODO_TECHDEBT(#543): Remove once the query clients start handling caching and invalidation.
+	sharedParams           *sharedtypes.Params
+	service                *sharedtypes.Service
+	serviceRelayDifficulty servicetypes.RelayMiningDifficulty
+}
+
+// ProxyRelayMeter is the offchain Supplier's rate limiter.
+// It ensures that no Application is over-serviced by the Supplier per session.
+// This is done by maintaining the max amount of stake the supplier can consume
+// per session and the amount of stake consumed by mined relays.
+type ProxyRelayMeter struct {
+	// supplierToAppMetricsMap is a map of supplier addresses to application address
+	// to the application's relay meter.
+	// Only known applications (i.e. have sent at least one relay) that have their stakes metered.
+	// This map gets reset every new session in order to meter new applications.
+	supplierToAppMetricsMap map[string]map[string]*appRelayMeter
+	// overServicingAllowanceCoins allows Suppliers to overservice applications.
+	// This entails providing a free service (i.e. mine for relays), that they will not be paid for onchain.
+	// This is common by some suppliers to build goodwill and receive a higher offchain quality-of-service rating.
+	// If negative, allow infinite overservicing.
+	// TODO_MAINNET(@red-0ne): Expose overServicingAllowanceCoins as a configuration parameter.
+	overServicingAllowanceCoins cosmostypes.Coin
+
+	// relayMeterMu ensures that relay meter operations are thread-safe.
+	relayMeterMu sync.Mutex
+
+	applicationQuerier client.ApplicationQueryClient
+	serviceQuerier     client.ServiceQueryClient
+	sharedQuerier      client.SharedQueryClient
+	eventsQueryClient  client.EventsQueryClient
+	blockQuerier       client.BlockClient
+}
```

Comment:
Rooting the meta the implementation:

The `RelayMiner` itself could be seen as a light client. Since it is already receiving Tx and NewBlock events through websockets (for multiple reasons), A block header verification is the only missing part to make them true light clients.

Even if the `RelayMiner` operators are very incentivized to have their own Pocket Network RPC node, we should expect seeing private setups where multiple `RelayMiner`s are sharing the same node.